### PR TITLE
[ASEditableTextNode] Maximum number of lines to display (#2777)

### PR DESCRIPTION
**Pull request by leotumwattana**
This adds the property . maximumLinesToDisplay to ASEditableTextNode which will limit the size calculation of the editableTextNode to the specified number of lines.

Calculation is done using TextKit's lineFragment to ensure height calculation is accurate even under mixed font conditions.

Addresses Issue: #2777 
Replaces: PR #2786 (which was based of branch 2.0)

### CODE REVIEW

**Comment by leotumwattana** [comment]
@Adlai-Holler: Here's the replacement PR based off Master. Thanks.

File: AsyncDisplayKit/ASEditableTextNode.mm [hunk 1]
```diff
@@ -313,6 +324,11 @@ - (UITextView *)textView
   return _textKitComponents.textView;
 }
 
+- (void)setMaximumLinesToDisplay:(NSUInteger)maximumLines
+{
+  _maximumLinesToDisplay = maximumLines;
```
Comment (Adlai-Holler) [inline AsyncDisplayKit/ASEditableTextNode.mm]: I believe we should add a `[self setNeedsLayout]` call here, to handle the unlikely case that the user changes this property on a live node.

File: AsyncDisplayKit/TextKit/ASTextKitComponents.mm [hunk 2]
```diff
@@ -66,4 +68,57 @@ - (CGSize)sizeForConstrainedWidth:(CGFloat)constrainedWidth
   return textSize;
 }
 
+- (CGSize)sizeForConstrainedWidth:(CGFloat)constrainedWidth
+              forMaxNumberOfLines:(NSInteger)maxNumberOfLines {
```
Comment (Adlai-Holler) [inline AsyncDisplayKit/TextKit/ASTextKitComponents.mm]: Open brace on the next line.

File: AsyncDisplayKit/TextKit/ASTextKitComponents.mm [hunk 3]
```diff
@@ -66,4 +68,57 @@ - (CGSize)sizeForConstrainedWidth:(CGFloat)constrainedWidth
...
+  CGRect rect = CGRectZero;
+  CGFloat height = 0;
+  CGFloat lastOriginY = -1.0;
+  NSUInteger numberOfLines = 0;
+  
+  glyphRange = [components.layoutManager glyphRangeForTextContainer:components.textContainer];
+  
+  while (lineRange.location < NSMaxRange(glyphRange)) {
+    
+    rect = [components.layoutManager lineFragmentRectForGlyphAtIndex:lineRange.location
```
Comment (Adlai-Holler) [inline AsyncDisplayKit/TextKit/ASTextKitComponents.mm]: Nit: remove blank lines at beginning and end of this `while`

File: AsyncDisplayKit/ASEditableTextNode.mm [hunk 4]
```diff
@@ -237,11 +237,22 @@ - (void)didLoad
...
+    textSize = [displayedComponents sizeForConstrainedWidth:constrainedSize.width
+                                        forMaxNumberOfLines: _maximumLinesToDisplay];
+  } else {
+    textSize = [displayedComponents sizeForConstrainedWidth:constrainedSize.width];
+  }
+  
   CGFloat width = std::ceil(textSize.width + _textContainerInset.left + _textContainerInset.right);
   CGFloat height = std::ceil(textSize.height + _textContainerInset.top + _textContainerInset.bottom);
   return CGSizeMake(std::fmin(width, constrainedSize.width), std::fmin(height, constrainedSize.height));
+
```
Comment (Adlai-Holler) [inline AsyncDisplayKit/ASEditableTextNode.mm]: Remove blank line

**Comment by leotumwattana** [comment]
Thanks @Adlai-Holler. Made the changes you mentioned above. Cheers. 👍

**Comment by Adlai-Holler** [comment]
Sorry for the delay here, time to land!

**Pull request merged**